src/mpid/ch4/shm/xpmem/xpmem_recv.h

```diff
@@ -9,71 +9,310 @@
+                slmt_cts_hdr->cnt_pool_addr, slmt_cts_hdr->dest_lrank, MPIDIG_REQUEST(rreq, rank));
+
+    /* Receiver sends CTS packet to sender */
+    mpi_errno =
+        MPIDI_SHM_do_ctrl_send(MPIDIG_REQUEST(rreq, rank), comm, MPIDI_SHM_XPMEM_SEND_LMT_CTS,
+                               &ctrl_hdr);
     MPIR_ERR_CHECK(mpi_errno);
 
-    MPIDI_Datatype_check_size(MPIDIG_REQUEST(rreq, datatype), MPIDIG_REQUEST(rreq, count), data_sz);
+    /* Data truncation checking */
+    recv_data_sz = MPL_MIN(src_data_sz, data_sz);
     if (src_data_sz > data_sz)
         rreq->status.MPI_ERROR = MPI_ERR_TRUNCATE;
 
-    /* Copy data to receive buffer */
-    recv_data_sz = MPL_MIN(src_data_sz, data_sz);
-    mpi_errno = MPIR_Localcopy(attached_sbuf, recv_data_sz,
-                               MPI_BYTE, (char *) MPIDIG_REQUEST(rreq, buffer),
-                               MPIDIG_REQUEST(rreq, count), MPIDIG_REQUEST(rreq, datatype));
+    /* Set receive status */
+    MPIR_STATUS_SET_COUNT(rreq->status, recv_data_sz);
+    rreq->status.MPI_SOURCE = MPIDIG_REQUEST(rreq, rank);
+    rreq->status.MPI_TAG = MPIDIG_REQUEST(rreq, tag);
+    dest_buf = MPIDIG_REQUEST(rreq, buffer);
+
+    mpi_errno =
+        MPIDI_XPMEM_seg_regist(src_lrank, src_data_sz, (void *) src_offset, &seg_ptr, &src_buf,
+                               MPIDI_XPMEM_USER_BUFFER);
+    MPIR_ERR_CHECK(mpi_errno);
 
-    XPMEM_TRACE("handle_lmt_recv: handle matched rreq %p [source %d, tag %d, context_id 0x%x],"
-                " copy dst %p, src %p, bytes %ld\n", rreq, MPIDIG_REQUEST(rreq, rank),
+    XPMEM_TRACE("handle_lmt_coop_recv: handle matched rreq %p [source %d, tag %d, context_id 0x%x],"
+                " copy dst %p, src %lx, bytes %ld\n", rreq, MPIDIG_REQUEST(rreq, rank),
                 MPIDIG_REQUEST(rreq, tag), MPIDIG_REQUEST(rreq, context_id),
-                (char *) MPIDIG_REQUEST(rreq, buffer), attached_sbuf, recv_data_sz);
+                (char *) MPIDIG_REQUEST(rreq, buffer), src_offset, recv_data_sz);
+    /* sender and receiver datatypes are both continuous, perform cooperative copy. */
+    mpi_errno =
+        MPIDI_XPMEM_do_lmt_coop_copy(src_buf, recv_data_sz,
+                                     (char *) dest_buf + dt_true_lb,
+                                     &MPIDI_XPMEM_REQUEST(rreq, counter_ptr)->counter, sreq_ptr,
+                                     rreq, &fin_type);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    if (fin_type != MPIDI_XPMEM_WAIT_FIN) {
+        if (fin_type == MPIDI_XPMEM_COPY_ALL_ISSUE_FIN) {
+            MPIDI_XPMEM_cnt_free_obj_t *cnt_obj =
+                MPL_malloc(sizeof(MPIDI_XPMEM_cnt_free_obj_t), MPL_MEM_OTHER);
+            MPIR_Assert(cnt_obj != NULL);
+            cnt_obj->counter_ptr = MPIDI_XPMEM_REQUEST(rreq, counter_ptr);
+            MPIDI_XPMEM_REQUEST(rreq, counter_ptr) = NULL;
+            DL_APPEND(MPIDI_XPMEM_global.cnt_free_queue, cnt_obj);
+        }
```

Comment:
I do not understand why you allocate a new counter when the copy finishes. Let us use the object API and I will read the code again.